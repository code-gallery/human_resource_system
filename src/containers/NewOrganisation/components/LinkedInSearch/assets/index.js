import linkedIn from './linkedin-signin.png'

export default {
  linkedIn
}
