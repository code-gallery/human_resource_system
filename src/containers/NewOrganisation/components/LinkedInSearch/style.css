.LinkedInSearch-signin {
  height: 45px;
  cursor: pointer;
}
