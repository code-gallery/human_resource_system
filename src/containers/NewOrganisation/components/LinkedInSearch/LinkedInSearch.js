import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import _isNil from 'lodash/isNil'
import _map from 'lodash/map'
import Form from 'components/Abstract/Form'
import LinkedInSearchResults from '../LinkedInSearchResults'
import Assets from './assets'
import './style.css'

class LinkedInSearch extends Form {
  constructor(props) {
    super(props)
    if (window.IN && window.IN.User) {
      window.IN.User.refresh()
    }
    this.state = {
      token: (window.IN && window.IN.ENV && window.IN.ENV.auth) ? window.IN.ENV.auth.oauth_token : '',
      showResults: false,
      pending: false,
      results: [],
      errors: []
    }
  }

  signIn = () => {
    window.IN.User.authorize(() => {
      this.setState({
        token: window.IN.ENV.auth.oauth_token
      })
    })
  }

  select = (item) => {
    const results = _map(this.state.results, (r) => {
      if (r.id === item.id) {
        r.selected = true
      }
      return r
    })

    this.setState({ results })
    this.props.save({ ...item })
  }

  search = (e) => {
    e.preventDefault()
    const { keywords, token } = this.state
    if (!_isNil(keywords) && keywords !== '') {
      this.setState({ showResults: false, pending: true })
      const fields = 'id,name,logo_url,website_url,specialties,employee-count-range,industries,twitter-id,founded-year'
      const searchBase = `company-search:(companies:(${fields}))`
      window.IN.API.Raw() // eslint-disable-line new-cap
        .url(`${searchBase}?keywords=${keywords}&oauth2_access_token=${token}&format=json&count=15`)
        .method('GET')
        .result((json) => {
          if (json.companies._total > 0) {
            this.setState({ results: json.companies.values, showResults: true, pending: false })
          } else {
            this.setState({ results: [], showResults: true, pending: false })
          }
        })
    } else {
      this.setState({
        errors: [ 'keywords' ]
      })
    }
  }

  render() {
    const { token, errors, results, showResults, pending } = this.state

    if (token === '') {
      return (
        <div className="LinkedInSearch">
          <p className="form-group">
            To use LinkedIn search, you first need to sign in and authorize APPII application.<br />
            The sign in form will open in a popup window. Please make sure your browser does not block it.
          </p>
          <p className="form-group">
            <img
              className="LinkedInSearch-signin"
              src={Assets.linkedIn}
              alt="Sign in with LinkedIn"
              onClick={this.signIn}
            />
          </p>
        </div>
      )
    }

    return (
      <div>
        <form>
          <h2 className="form-group">
            Search for an organisation
          </h2>
          <div
            className={classNames('form-group', {
              'has-error': errors.indexOf('keywords') !== -1
            })}
          >
            <input
              placeholder="Enter a name"
              name="keywords"
              type="text"
              onChange={this.onChangeInput}
            />
          </div>
          <div className="form-group">
            <button
              className="form-btn save"
              onClick={this.search}
              disabled={pending}
            >
              {(pending) ? 'Searching' : 'Search'}
            </button>
          </div>
        </form>
        {showResults &&
          <LinkedInSearchResults results={results} select={this.select} />
        }
      </div>
    )
  }
}

LinkedInSearch.propTypes = {
  save: PropTypes.func.isRequired
}

export default LinkedInSearch
