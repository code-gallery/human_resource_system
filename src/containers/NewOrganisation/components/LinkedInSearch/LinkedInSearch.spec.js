import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import LinkedInSearch from './LinkedInSearch'

describe('containers/NewOrganisation/components/LinkedInSearch', () => {
  beforeEach(() => {
    window.IN = {}
    window.IN.User = {
      refresh: jest.fn(),
      authorize: function(cb) {
        cb()
      }
    }
  })

  describe('when no linkedIn token', () => {
    it('renders correctly', () => {
      window.IN.ENV = {
        auth: { oauth_token: '' }
      }
      const props = {
        save: jest.fn()
      }
      const tree = renderer.create(
        <LinkedInSearch {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('when there is linkedIn token', () => {
    it('renders correctly', () => {
      window.IN.ENV = {
        auth: { oauth_token: 'token_value_here' }
      }
      const props = {
        save: jest.fn()
      }
      const tree = renderer.create(
        <LinkedInSearch {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('<LinkedInSearch />', () => {
    let props
    let wrapper
    beforeEach(() => {
      props = {
        save: jest.fn()
      }
      window.IN.ENV = {
        auth: { oauth_token: '' }
      }
      wrapper = shallow(
        <LinkedInSearch {...props} />
      )
    })

    describe('signIn', () => {
      it('sets states correctly', () => {
        window.IN.ENV = {
          auth: { oauth_token: 'token_value_here' }
        }
        wrapper.setState({ token: null })
        wrapper.instance().signIn()
        expect(wrapper.state('token')).toEqual('token_value_here')
      })
    })

    describe('select', () => {
      it('calls props.save and update state correctly', () => {
        wrapper.setState({
          results: [
            { id: 12, name: 'ab' },
            { id: 123, name: 'abc' }
          ]
        })
        wrapper.instance().select({ id: 123, name: 'abc' })
        expect(wrapper.state('results')[1]).toEqual({ id: 123, name: 'abc', selected: true })
        expect(props.save).toHaveBeenCalledWith({ id: 123, name: 'abc' })
      })
    })

    describe('search', () => {
      describe('when keywords is invalid', () => {
        it('updates state correctly', () => {
          expect(wrapper.state('errors')).toEqual([])
          wrapper.instance().search({ preventDefault: jest.fn() })
          expect(wrapper.state('errors')).toEqual([ 'keywords' ])
        })
      })

      describe('when keywords is valid', () => {
        describe('when there are results', () => {
          it('updates state correctly', () => {
            const values = [
              { id: 1, name: 'ab' },
              { id: 2, name: 'abc' },
              { id: 3, name: 'abcd' }
            ]
            window.IN.API = {}
            window.IN.API.Raw = function() {
              return {
                url: function() {
                  return {
                    method: function() {
                      return {
                        result: function(cb) {
                          return cb({
                            companies: {
                              _total: 3,
                              values
                            }
                          })
                        }
                      }
                    }
                  }
                }
              }
            }
            wrapper.setState({ keywords: 'Apple', token: 'token' })
            wrapper.instance().search({ preventDefault: jest.fn() })
            expect(wrapper.state('pending')).toEqual(false)
            expect(wrapper.state('showResults')).toEqual(true)
            expect(wrapper.state('results')).toEqual(values)
          })
        })

        describe('when there are no results', () => {
          it('updates state correctly', () => {
            const values = []
            window.IN.API = {}
            window.IN.API.Raw = function() {
              return {
                url: function() {
                  return {
                    method: function() {
                      return {
                        result: function(cb) {
                          return cb({
                            companies: {
                              _total: 0,
                              values
                            }
                          })
                        }
                      }
                    }
                  }
                }
              }
            }
            wrapper.setState({ keywords: 'wrerer', token: 'token' })
            wrapper.instance().search({ preventDefault: jest.fn() })
            expect(wrapper.state('pending')).toEqual(false)
            expect(wrapper.state('showResults')).toEqual(true)
            expect(wrapper.state('results')).toEqual(values)
          })
        })
      })
    })
  })
})
