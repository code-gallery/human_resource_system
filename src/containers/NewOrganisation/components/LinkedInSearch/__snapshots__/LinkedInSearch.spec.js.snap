// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/NewOrganisation/components/LinkedInSearch when no linkedIn token renders correctly 1`] = `
<div
  className="LinkedInSearch"
>
  <p
    className="form-group"
  >
    To use LinkedIn search, you first need to sign in and authorize APPII application.
    <br />
    The sign in form will open in a popup window. Please make sure your browser does not block it.
  </p>
  <p
    className="form-group"
  >
    <img
      alt="Sign in with LinkedIn"
      className="LinkedInSearch-signin"
      onClick={[Function]}
      src="linkedin-signin.png"
    />
  </p>
</div>
`;

exports[`containers/NewOrganisation/components/LinkedInSearch when there is linkedIn token renders correctly 1`] = `
<div>
  <form>
    <h2
      className="form-group"
    >
      Search for an organisation
    </h2>
    <div
      className="form-group"
    >
      <input
        name="keywords"
        onChange={[Function]}
        placeholder="Enter a name"
        type="text"
      />
    </div>
    <div
      className="form-group"
    >
      <button
        className="form-btn save"
        disabled={false}
        onClick={[Function]}
      >
        Search
      </button>
    </div>
  </form>
</div>
`;
